src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -105,62 +108,180 @@ public void run() {
+                  TAG,
+                  "Failed to resolve service '" + serviceInfo.getServiceName() + "': " + errorCode);
+              chipMdnsCallback.handleServiceResolve(
+                  serviceInfo.getServiceName(),
+                  serviceInfo.getServiceType(),
+                  null,
+                  null,
+                  0,
+                  null,
+                  callbackHandle,
+                  contextHandle);
+
+              if (multicastLock.isHeld()) {
+                multicastLock.release();
+
+                if (nsdManagerResolverAvailState != null) {
+                  nsdManagerResolverAvailState.signalFree();
+                }
               }
+              mainThreadHandler.removeCallbacks(timeoutRunnable);
             }
-            mainThreadHandler.removeCallbacks(timeoutRunnable);
-          }
 
-          @Override
-          public void onServiceResolved(NsdServiceInfo serviceInfo) {
-            Log.i(
-                TAG,
-                "Resolved service '"
-                    + serviceInfo.getServiceName()
-                    + "' to "
-                    + serviceInfo.getHost());
-            // TODO: Find out if DNS-SD results for Android should contain interface ID
-            chipMdnsCallback.handleServiceResolve(
-                instanceName,
-                serviceType,
-                serviceInfo.getHost().getHostName(),
-                serviceInfo.getHost().getHostAddress(),
-                serviceInfo.getPort(),
-                serviceInfo.getAttributes(),
-                callbackHandle,
-                contextHandle);
+            @Override
+            public void onServiceResolved(NsdServiceInfo serviceInfo) {
+              Log.i(
+                  TAG,
+                  "Resolved service '"
+                      + serviceInfo.getServiceName()
+                      + "' to "
+                      + serviceInfo.getHost());
+              // TODO: Find out if DNS-SD results for Android should contain interface ID
+              chipMdnsCallback.handleServiceResolve(
+                  serviceInfo.getServiceName(),
+                  serviceInfo.getServiceType(),
+                  serviceInfo.getHost().getHostName(),
+                  serviceInfo.getHost().getHostAddress(),
+                  serviceInfo.getPort(),
+                  serviceInfo.getAttributes(),
+                  callbackHandle,
+                  contextHandle);
 
-            if (multicastLock.isHeld()) {
-              multicastLock.release();
+              if (multicastLock.isHeld()) {
+                multicastLock.release();
 
-              if (nsdManagerResolverAvailState != null) {
-                nsdManagerResolverAvailState.signalFree();
+                if (nsdManagerResolverAvailState != null) {
+                  nsdManagerResolverAvailState.signalFree();
+                }
               }
+              mainThreadHandler.removeCallbacks(timeoutRunnable);
             }
```

Comment:
nit, this code is repeated, would be cleaner in a separate method

Reply:
Pre-existing code. I'll refactor in a separate PR - same as the acquireResolver() change above.